src/core/Ekubo.sol

```diff
@@ -0,0 +1,459 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.25;
+
+import {IERC20} from "@forge-std/interfaces/IERC20.sol";
+import {ISignatureTransfer} from "@permit2/interfaces/ISignatureTransfer.sol";
+import {SafeTransferLib} from "../vendor/SafeTransferLib.sol";
+import {SettlerAbstract} from "../SettlerAbstract.sol";
+import {Ternary} from "../utils/Ternary.sol";
+import {UnsafeMath} from "../utils/UnsafeMath.sol";
+import {Panic} from "../utils/Panic.sol";
+
+import {TooMuchSlippage, ZeroSellAmount} from "./SettlerErrors.sol";
+
+import {CreditDebt, Encoder, NotesLib, StateLib, Decoder, Take} from "./FlashAccountingCommon.sol";
+
+type Config is bytes32;
+
+type SqrtRatio is uint96;
+
+error AmountSpecifiedTooLarge(int256 amount);
+
+// Each pool has its own state associated with this key
+struct PoolKey {
+    address token0;
+    address token1;
+    Config config;
+}
+
+interface IEkuboCore {
+    // The entrypoint for all operations on the core contract
+    function lock() external;
+
+    // Swap tokens
+    function swap_611415377(
+        PoolKey memory poolKey,
+        int128 amount,
+        bool isToken1,
+        SqrtRatio sqrtRatioLimit,
+        uint256 skipAhead
+    ) external payable returns (int128 delta0, int128 delta1);
+
+    // Pay for swapped tokens
+    function pay(address token) external returns (uint128 payment);
+
+    // Get swapped tokens
+    function withdraw(address token, address recipient, uint128 amount) external;
+}
+
+IEkuboCore constant CORE = IEkuboCore(0xe0e0e08A6A4b9Dc7bD67BCB7aadE5cF48157d444);
+
+/// @notice Interface for the callback executed when an address locks core
+interface IEkuboCallbacks {
+    /// @notice Called by Core on `msg.sender` when a lock is acquired
+    /// @param id The id assigned to the action
+    /// @return Any data that you want to be returned from the lock call
+    function locked(uint256 id) external returns (bytes memory);
+
+    /// @notice Called by Core on `msg.sender` to collect assets
+    /// @param id The id assigned to the action
+    /// @param token The token to pay on
+    function payCallback(uint256 id, address token) external;
+}
+
+library UnsafeEkuboCore {
+    function unsafeSwap(
+        IEkuboCore core,
+        PoolKey memory poolKey,
+        int128 amount,
+        bool isToken1,
+        SqrtRatio sqrtRatioLimit,
+        uint256 skipAhead
+    ) internal returns (int128 delta0, int128 delta1) {
+        assembly ("memory-safe") {
+            let ptr := mload(0x40)
+
+            // mstore(ptr, 0x00000000) // selector for `swap_611415377((address,address,bytes32),int128,bool,uint96,uint256)`
+            // TODO: Check that native ETH is handled properly (0x0 vs 0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee)
+            mcopy(add(0x20, ptr), key, 0x60)
+            mstore(add(0x80, ptr), amount)
```

Comment:
needs `SIGNEXTEND`

Reply:
not going to miss this one next time, sorry